riscv_isac/coverage.py

```diff
@@ -957,17 +963,92 @@ def old_fn_csr_comb_covpt(csr_reg):
             for i in csr_regfile.csr_regs:
                 instr_vars[i] = int(csr_regfile[i],16)
 
+            instr.iptw_update(instr_vars, iptw_dict)
+            instr.ptw_update(instr_vars)
+
+            for i in iptw_dict:
+                instr_vars[i] = (iptw_dict[i])
+
             csr_write_vals = {}
             if instr.csr_commit is not None:
                 for commit in instr.csr_commit:
-                    if commit[0] == "CSR" and commit[3]:
-                        csr_write_vals[commit[1]] = int(commit[3],16)
+                    if commit[0] == "CSR" and commit[4] and commit[2] == '<-':
+                        csr_write_vals[commit[1]] = int(commit[4],16)
             def write_fn_csr_comb_covpt(csr_reg):
                 if csr_reg in csr_write_vals:
                     return csr_write_vals[csr_reg]
                 else:
                     return int(csr_regfile[csr_reg],16)
+                
+            csr_read_vals = {}
+            if instr.csr_commit is not None:
+                for csr_commit in instr.csr_commit:
+                    if(csr_commit[0] == "CSR") and (csr_commit[3]) and (csr_commit[2] == '->'):
+                        csr_read_vals[csr_commit[1]] = int(csr_commit[3],16)
+            def read_fn_csr_comb_covpt(csr_reg):
```

Comment:
I think this function is redundant. The value for the csr being read is the same as the value in the arch state. Hence the coverpoints should not look different and follow the same format as `mcause == 1`. Something worth doing would be to check the value of the csr in the arch_state dictionary and verify it is correct.

Reply:
This function is not redundant as the arch_state dictionary only takes the values that are written in the log. So, we faced a problem like for the `mcause` we did not know whether there was an exception or not because this register was never written in the log file with the exception. Therefore, to keep track of the values that were only read in the log which is differentiated using the `<-` or `->`, I had used this function.

So, earlier, the c_sail.py used to track only the registers that were written/overwritten in the log

https://github.com/riscv-software-src/riscv-isac/blob/db7504037ee611aa5b7e0b40e5207ae4871ef574/riscv_isac/plugins/c_sail.py#L20

But, now it takes care of both the registers that are read and also written

```python
    instr_pattern_c_sail_csr_reg_val = re.compile('(?P<CSR>CSR|clint::tick)\s(?P<reg>[a-z0-9]+)\s(.*?)\s(?P<val>[0-9xABCDEF]+)(?:\s\(input:\s(?P<input_val>[0-9xABCDEF]+)\))?')
```

Reply:
In the absence of Sail explicitly generating a message for each update to architectural state (e.g. all the trap registers) we need something like this.
I have filed an issue with the Sail team to ensure that all implicit writes of CSRs appear in the Sail output (we will want some qualifications, e.g. minstret or mcycle should default to not logged, and maybe there are some categories we can selectively enable/disable

Reply:
And to be explicit: until the Sail team ensures that all implicit writes of CSRs appear in the Sail output , we need to do this

Reply:
This is a stop-gap solution which ultimately introduces more issues because the meaning of the coverpoints get changed. This means that the coverpoints will have to be re-written to not include the `read_fn_csr_comb_covpt`. The primary fix would be to fix it in SAIL. A secondary fix would be to explictly overwrite the value being read from the log(indicated with `<-`) before evaluating and printing a warning that this feature would be deprecated in the future. Alternatively, just a warning to indicate that the value being read and the value in the logs are different. This check should stay on as a feature IMO.

Reply:
Actually, there is a seperate track for the read_fn_csr_comb which is done by `csr_read_vals` dictionary, which is something like for a instruction lets say:
```
mem[X,0x80000A78] -> 0x23F3
mem[X,0x80000A7A] -> 0x3410
[202] [M]: 0x80000A78 (0x341023F3) csrrs t2, mepc, zero
CSR mepc -> 0x800003C0
x7 <- 0x800003C0
```
The value in the read_csr will be `csr_read_vals = {'mepc': 2147484608}` and this value is never tracked by either the arch_state nor the csr_regfile because it was never written by the user rather by the hart, so this can be accessed using the coverpoint:
```
csr_comb:
   read_csr("mepc") == 2147484608 : 0
```
This is just for an example. The main purpose for this functionality was to check the values for the mcause to check whether specific exception occured in the test or not as it will not be tracked by the arch_state or csr_regfile, but I have updated that as well(please refer to https://github.com/riscv-software-src/riscv-isac/pull/80#issuecomment-1910112010), so if you want me to remove this feature, please let me know.

Reply:
This feature has been removed.